### PR TITLE
Number, Boolean and String prototypes revert to ES5 semantics

### PR DESCRIPTION
**Pull request by littledan**
At the August 2015 TC39 meeting, we came to consensus that, for web
compatibility reasons, the prototype of Number, Boolean and String
needed to act like in ES5, with [[NumberData]] of +0, etc. This
patch reverts ECMAScript to have those semantics.

### CODE REVIEW

**Comment by rwaldron** [comment]
Nice :+1:

File: spec.html [hunk 1]
```diff
@@ -23527,7 +23527,8 @@
     <!-- es6num="19.3.3" -->
     <emu-clause id="sec-properties-of-the-boolean-prototype-object">
       <h1>Properties of the Boolean Prototype Object</h1>
-      <p>The Boolean prototype object is the intrinsic object %BooleanPrototype%. The Boolean prototype object is an ordinary object. It is not a Boolean instance and does not have a [[BooleanData]] internal slot.</p>
```
Comment (anba) [inline spec.html]: Why did you remove the sentence "The Boolean prototype object is an ordinary object."? That part is still correct, even if `Boolean.prototype` gets a [[BooleanData]] internal slot.

Comment (rwaldron) [inline spec.html]: I believe it would be a Boolean object, not ordinary object?

(from ES5)
> The Boolean prototype object is itself a Boolean object (its [[Class]] is "Boolean") whose value is false.

Comment (ljharb) [inline spec.html]: Exactly, as I understand it, ordinary objects don't have internal slots.

Comment (rwaldron) [inline spec.html]: I think confusion is: 

> The Boolean prototype object is the intrinsic object %BooleanPrototype%. The Boolean prototype object is an ordinary object

Seems like it's saying contradictory things, but isn't.

I'd recommend a complete change to: 

"The Boolean prototype object is itself a Boolean object, it has a [[BooleanData]] internal slot with the value false."

Comment (anba) [inline spec.html]: Objects which use the ordinary internal methods from [9.1](http://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots) are called ["Ordinary objects" (4.3.6)](http://tc39.github.io/ecma262/#sec-ordinary-object). This does not mean ordinary objects are not allowed to have slots, otherwise it wouldn't make sense to have an [ObjectCreate (9.1.13)](http://tc39.github.io/ecma262/#sec-objectcreate) method to create ordinary objects with custom slots.

Comment (bterlson) [inline spec.html]: [Ordinary objects have many internal slots](http://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots), so I think @anba is correct. There needs to be some verbage here about what kind of object the thing actually is. I think @rwaldron's suggestion is best - apply to all types below and then I can pull this in :)

Comment (anba) [inline spec.html]: > "The Boolean prototype object is itself a Boolean object, it has a [[BooleanData]] internal slot with the value false."

That definition is okay, because Boolean objects are also ordinary objects. It's just important to specify that `Boolean.prototype` is an ordinary object, otherwise implementations could redefine one of the internal methods of `Boolean.prototype`.

Comment (domenic) [inline spec.html]: Nit on @rwaldron's proposed verbiage: it is a run on sentence, and should either be split into two sentences or use a semicolon instead of a comma.

Comment (bterlson) [inline spec.html]: I vote for semicolon! It's not every day you get to use one! :confetti_ball:

Comment (littledan) [inline spec.html]: Yay semicolons! I'm all for that.

OK, I agree with the ordinary object description (it's not an exotic object). However, I don't really know exactly what it means to say that it "is a Boolean object". As far as I can tell, what makes something a Boolean object is that it has the [[BooleanValue]] internal slot, so the sentence would seem a little redundant.

However, I don't understand the concern about implementations having the right to redefine internal methods for Boolean.prototype just because it's an exotic object. Do implementations have this right on, say, Arrays?

**Comment by littledan** [comment]
OK, I added an additional patch indicating that they are ordinary objects. What do you think?

**Comment by zloirock** [comment]
Why only `Number`, `Boolean` and `String`? For example, [xregexp 2](https://github.com/slevithan/xregexp/blob/v2.0.0/xregexp-all.js#L82).

**Comment by littledan** [comment]
Fortunately, the sticky flag actually is on the prototype in ES2015, so that code should work in the future.

It's important for RegExp to be an ordinary object *without* that internal slot, because RegExp.prototype.compile should not work on it, otherwise RegExp would be stateful and not meet SES constraints.

**Comment by anba** [comment]
> Fortunately, the sticky flag actually is on the prototype in ES2015, so that code should work in the future.

Nope, it throws a `TypeError` at runtime.

File: spec.html (hunk 1 above)
Comment (anba) [inline spec.html]: > However, I don't really know exactly what it means to say that it "is a Boolean object". As far as I can tell, what makes something a Boolean object is that it has the [[BooleanValue]] internal slot, so the sentence would seem a little redundant.

In that case you could also write:
"The Boolean prototype object is itself a Boolean object; its [[BooleanData]] internal slot has the value false."


> However, I don't understand the concern about implementations having the right to redefine internal methods for Boolean.prototype just because it's an exotic object. Do implementations have this right on, say, Arrays?

Exotic objects have a different definition for one (or more) of the internal methods (cf. [4.3.7](http://tc39.github.io/ecma262/#sec-exotic-object)). The term "Array exotic object" has a specific meaning, it refers to the definition in [9.4.2](http://tc39.github.io/ecma262/#sec-array-exotic-objects) which only specifies to use a different implementation for [[DefineOwnProperty]].

**Comment by anba** [comment]
The new `RegExp.prototype` accessors (`global`, `ignoreCase`, `multiline`, `sticky` and `unicode`) could also be changed to return false if given a non-RegExp object. Basically that means using a similar approach like the web-compat workaround in `Date.prototype.toString`.

**Comment by littledan** [comment]
anba, Oh right. Well, that seems like a reasonable idea (if this library is widely used and it seems like we won't be able to get all uses updated to using hasOwnProperty in time). Should we make this change?

**Comment by littledan** [comment]
Switched to the wording suggested by @anba

**Comment by bterlson** [comment]
Alright this seems good. Pulling it in!

**Comment by bterlson** [comment]
Merged in a3ced3c.

File: spec.html (hunk 1 above)
Comment (allenwb) [inline spec.html]: the term, "Boolean object" is defined in [4.3.16](http://ecma-international.org/ecma-262/6.0/#sec-boolean-object).

Generally, in the ES<=6 specs, a phrase like "a \<construct name> object" means the same thing as "a \<constructor name> instance".  In other words, an instance of the named constructor.
